File: modules/core/app/model/exercisereading/ReadingResult.java
package model.exercisereading;

import java.util.List;

import model.WithId;

public class ReadingResult<T extends WithId> extends AbstractReadingResult {

  private List<T> read;

  public ReadingResult(String theJson, String theJsonSchema, List<T> theRead) {
    super(theJson, theJsonSchema);
    read = theRead;
  }

  public List<T> getRead() {
    return read;
  }

  @Override
  public boolean isSuccess() {
    return true;
  }

}

File: modules/core/app/model/exercisereading/ReadingResult.scala
package model.exercisereading

import java.nio.file.Path

import com.github.fge.jsonschema.core.report.ProcessingReport
import model.JsonReadable

import scala.collection.JavaConverters.asScalaIteratorConverter
import scala.util.Try


case class SingleReadingResult[R <: JsonReadable](read: R, var fileResults: List[Try[Path]] = List.empty)


sealed abstract class AbstractReadingResult

case class ReadingResult[R <: JsonReadable](read: List[SingleReadingResult[R]]) extends AbstractReadingResult

case class ReadingError(json: String, jsonSchema: String, report: ProcessingReport) extends AbstractReadingResult {
  def getCauses: Iterator[String] = report.iterator.asScala.map(_.toString)
}

case class ReadingFailure(error: Throwable) extends AbstractReadingResult